### PR TITLE
Support mapping to fields

### PR DESCRIPTION
**Pull request by ajcvickers**
Issues #4357, #4855

This change adds fluent and lower-level APIs to set backing fields or properties with no fields as well as conventions for using the fields and a mechanism to provide control over when the property is used and when the field is used.

Also see #6266, #6267, and #6268.

/cc @rowanmiller for API and docs

### CODE REVIEW

**Comment by anpete** [comment]
@ajcvickers Any guidance on specific areas of this to review?

**Comment by ajcvickers** [comment]
@anpete Probably the biggest change is the way fields are found and set in metadata when using the convention or the API. Then possibly the rules for choosing a field or a property based on mode.

**Comment by anpete** [comment]
:shipit:

File: src/Microsoft.EntityFrameworkCore/Metadata/Internal/PropertyBaseExtensions.cs [hunk 1]
```diff
@@ -145,6 +147,179 @@ public static PropertyInfo GetPropertyInfo([NotNull] this IPropertyBase property
...
+
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used 
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        public static object FindMemberInfo(
+            [NotNull] this IPropertyBase propertyBase,
+            bool forConstruction,
+            bool forSet)
+        {
```
Comment (AndriySvyryd) [inline src/Microsoft.EntityFrameworkCore/Metadata/Internal/PropertyBaseExtensions.cs]: Consider using an out parameter or return a Tuple

File: src/Microsoft.EntityFrameworkCore/Metadata/Internal/InternalRelationshipBuilder.cs [hunk 2]
```diff
@@ -1153,9 +1153,7 @@ private bool CanSetUnique(bool unique, ConfigurationSource? configurationSource,
                     && Metadata.DeclaringEntityType == dependentEntityType)
                 {
                     Debug.Assert(!shouldResetToPrincipal
-                                 && !shouldResetToDependent
-                                 && !shouldResetPrincipalProperties
-                                 && !shouldResetDependentProperties);
```
Comment (AndriySvyryd) [inline src/Microsoft.EntityFrameworkCore/Metadata/Internal/InternalRelationshipBuilder.cs]: Note for myself to revision this assert

File: src/Microsoft.EntityFrameworkCore/Metadata/Conventions/Internal/BackingFieldConvention.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        }
+
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        protected virtual FieldInfo TryMatchFieldName(
+            [NotNull] Type entityType, [NotNull] Type propertyType, [NotNull] string propertyName)
+        {
+            var fields = entityType.GetRuntimeFields().ToDictionary(f => f.Name);
```
Comment (AndriySvyryd) [inline src/Microsoft.EntityFrameworkCore/Metadata/Conventions/Internal/BackingFieldConvention.cs]: We'll likely need to cache this

**Comment by AndriySvyryd** [comment]
:shipit:

**Pull request merged**